Zend/zend_alloc.c

```diff
@@ -329,16 +347,25 @@ struct _zend_mm_heap {
 	double             avg_chunks_count;		/* average number of chunks allocated per request */
 	int                last_chunks_delete_boundary; /* number of chunks after last deletion */
 	int                last_chunks_delete_count;    /* number of deletion over the last boundary */
+#if ZEND_MM_CUSTOM
+	HashTable *tracked_allocs;
+#endif
+	pid_t pid;
+	zend_random_bytes_insecure_state rand_state;
+};
+
+/* This contains security-sensitive data, and is thus mapped as read-only at run-time right after the _zend_mm_heap struct
+ * and accessed via the ZEND_MM_RO_HEAP macro.*/
+struct _zend_mm_ro_heap {
+	uintptr_t          shadow_key;              /* free slot shadow ptr xor key */
```

Comment:
Moving `shadow_key` to ro_heap will possibly add some non-negligible overhead as the field is fetched in the fast path, and it has to be fetched from a non-constant offset now (as `REAL_PAGE_SIZE` is non-constant).

I'm not sure it's worth it to write-protect this field. My reasoning is that if the attacker can override a field in `zend_mm_heap` (such as `shadow_key`), we can assume that the protection offered by `shadow_key` has already been bypassed, and overriding the freelist is either already done, or of no interest at this point? But maybe you have a better reasoning.